packages/front-end/pages/integrations/vercel/index.tsx

```diff
@@ -0,0 +1,145 @@
+import { useRouter } from "next/router";
+import { useEffect, useState } from "react";
+import { useAuth } from "../../../services/auth";
+import { Environment } from "back-end/types/organization";
+import { GbVercelEnvMap } from "back-end/types/vercel";
+import EnvironmentModal from "../../../components/Settings/EnvironmentModal";
+import SelectField from "../../../components/Forms/SelectField";
+import useOrgSettings from "../../../hooks/useOrgSettings";
+import Modal from "../../../components/Modal";
+import useApi from "../../../hooks/useApi";
+
+export default function VercelIntegrationPage() {
+  const router = useRouter();
+  const { code, configurationId, teamId, next } = router.query;
+
+  const { apiCall } = useAuth();
+  const { environments } = useOrgSettings();
+
+  const { data } = useApi<{ hasToken: boolean }>("/vercel/has-token");
+  const [integrationAlreadyExists, setIntegrationAlreadyExists] = useState(
+    false
+  );
+
+  const [envModalOpen, setEnvModalOpen] = useState<Partial<Environment> | null>(
+    null
+  );
+  const [gbVercelEnvMap, setGbVercelEnvMap] = useState<GbVercelEnvMap>([
+    { vercel: "production", gb: "" },
+    { vercel: "preview", gb: "" },
+    { vercel: "development", gb: "" },
+  ]);
+
+  useEffect(() => {
+    if (data?.hasToken !== undefined) {
+      if (data.hasToken) return setIntegrationAlreadyExists(true);
+      postToken();
+    }
+  }, [data]);
+
+  async function postToken() {
+    const options = {
+      method: "POST",
+      body: JSON.stringify({
+        code,
+        configurationId,
+        teamId: teamId ? teamId : null,
+      }),
+    };
+    apiCall("/vercel/token", options).catch(() => {
+      //do nothing
+    });
+  }
+
+  return (
+    <>
+      {integrationAlreadyExists ? (
+        <Modal
+          open
+          close={() => close()}
+          cta="Continue"
+          submit={async () => {
+            setIntegrationAlreadyExists(false);
+            postToken();
+          }}
+          autoCloseOnSubmit={false}
+        >
+          <div className="alert alert-warning">
+            <strong>Notice:</strong> A Vercel integration already exists for
+            your organization. By clicking <strong>{`"Continue"`}</strong> you
+            will overwrite the existing integration. Click{" "}
+            <strong>{`"Cancel"`}</strong> to avoid your Vercel integration being
+            overwritten.
+          </div>
+        </Modal>
+      ) : (
+        <>
+          {envModalOpen && (
+            <EnvironmentModal
+              existing={envModalOpen}
+              close={() => setEnvModalOpen(null)}
+              onSuccess={() => setIntegrationAlreadyExists(false)}
+            />
+          )}
+          <Modal
```

Comment:
Is there any reason to render this modal at all if `envModalOpen` is true?  You could do this instead:

```tsx
{envModalOpen 
  ? <EnvironmentModal/>
  : <Modal...>...</Modal>
}
```

Another option is to set the `open` prop to `false` when the environment modal is open.  Either way, this will remove the need for the `showCTAs` setting.